src/main/java/uk/gov/justice/hmpps/prison/service/enteringandleaving/DischargeToHospitalService.kt

```diff
@@ -0,0 +1,119 @@
+package uk.gov.justice.hmpps.prison.service.enteringandleaving
+
+import org.slf4j.LoggerFactory
+import org.springframework.data.repository.findByIdOrNull
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+import uk.gov.justice.hmpps.prison.api.model.InmateDetail
+import uk.gov.justice.hmpps.prison.api.model.RequestForNewBooking
+import uk.gov.justice.hmpps.prison.api.model.RequestToDischargePrisoner
+import uk.gov.justice.hmpps.prison.api.model.RequestToReleasePrisoner
+import uk.gov.justice.hmpps.prison.repository.jpa.model.AgencyLocation
+import uk.gov.justice.hmpps.prison.repository.jpa.model.ExternalMovement
+import uk.gov.justice.hmpps.prison.repository.jpa.model.MovementReason
+import uk.gov.justice.hmpps.prison.repository.jpa.model.MovementType
+import uk.gov.justice.hmpps.prison.repository.jpa.model.Offender
+import uk.gov.justice.hmpps.prison.repository.jpa.model.OffenderBooking
+import uk.gov.justice.hmpps.prison.repository.jpa.repository.AgencyLocationRepository
+import uk.gov.justice.hmpps.prison.repository.jpa.repository.OffenderBookingRepository
+import uk.gov.justice.hmpps.prison.repository.jpa.repository.OffenderRepository
+import uk.gov.justice.hmpps.prison.service.BadRequestException
+import uk.gov.justice.hmpps.prison.service.EntityNotFoundException
+import uk.gov.justice.hmpps.prison.service.PrisonerTransferService
+import uk.gov.justice.hmpps.prison.service.transformers.OffenderTransformer
+import java.time.LocalDateTime
+import kotlin.Result.Companion.failure
+import kotlin.Result.Companion.success
+import kotlin.math.min
+
+@Service
+@Transactional
+class DischargeToHospitalService(
+  private val offenderRepository: OffenderRepository,
+  private val offenderBookingRepository: OffenderBookingRepository,
+  private val agencyLocationRepository: AgencyLocationRepository,
+  private val bookingIntoPrisonService: BookingIntoPrisonService,
+  private val externalMovementService: ExternalMovementService,
+  private val releasePrisonerService: ReleasePrisonerService,
+  private val transformer: OffenderTransformer,
+) {
+  private val log = LoggerFactory.getLogger(this::class.java)
+
+  fun dischargeToHospital(offenderNo: String, request: RequestToDischargePrisoner): InmateDetail {
+    val dischargeTime = request.dischargeTime ?: LocalDateTime.now()
+    val toLocation = findHospitalLocation(request.hospitalLocationCode).getOrThrow()
+    val offenderBooking = getOrCreateBooking(offenderNo, dischargeTime, request.fromLocationId, request.supportingPrisonId).getOrThrow()
+    val lastMovement = offenderBooking.lastMovement.orElse(null)
+
+    return if (!lastMovement.isRelease()) {
```

Comment:
how come we don't get a compile error given lastMovement can be null?

Reply:
Ah, I see below, a function on a nullable movement 👍

Reply:
Because the extension function is on the nullable type:

`private fun ExternalMovement?.isRelease() = this?.movementType?.code == MovementType.REL.code`